Server/mods/deathmatch/logic/CObjectSync.cpp

```diff
@@ -196,43 +193,106 @@ void CObjectSync::Packet_ObjectSync(CObjectSyncPacket& Packet)
+                CElement* pAttacker = CElementIDs::GetElement(pData->attackerID);
 
-                // Is the player syncing this object?
-                if ((pObject->GetSyncer() == pPlayer) && pObject->CanUpdateSync(pData->ucSyncTimeContext))
+                if (pData->fHealth > 0)
                 {
-                    // Apply the data to the object
-                    if (pData->ucFlags & 0x1)
+                    float fLoss = fPreviousHealth - pData->fHealth;
+
+                    if (fLoss > 0)
                     {
-                        pObject->SetPosition(pData->vecPosition);
-                        g_pGame->GetColManager()->DoHitDetection(pObject->GetPosition(), pObject);
+                        CLuaArguments Arguments;
+                        Arguments.PushNumber(fLoss);
+
+                        if (pAttacker)
+                            Arguments.PushElement(pAttacker);
+                        else
+                            Arguments.PushNil();
+
+                        pObject->CallEvent("onObjectDamage", Arguments);
                     }
-                    if (pData->ucFlags & 0x2)
-                        pObject->SetRotation(pData->vecRotation);
-                    if (pData->ucFlags & 0x4)
-                        pObject->SetHealth(pData->fHealth);
+                }
+                else // object has been break?
+                {
+                    CLuaArguments Arguments;
+                    if (pAttacker)
+                        Arguments.PushElement(pAttacker);
+                    else
+                        Arguments.PushNil();
 
-                    // Send this sync
-                    pData->bSend = true;
+                    pObject->CallEvent("onObjectBreak", Arguments);
                 }
             }
         }
+        if (pData->ucFlags & 0x20)            // Sync inWater state from client
+            pObject->SetInWater(pData->bIsInWater);
+        // Sync properties
+        if (pData->ucFlags & 0x40)
+            pObject->SetMass(pData->fMass);
+        if (pData->ucFlags & 0x80)
+            pObject->SetTurnMass(pData->fTurnMass);
+        if (pData->ucFlags & 0x100)
+            pObject->SetAirResistance(pData->fAirResistance);
+        if (pData->ucFlags & 0x200)
+            pObject->SetElasticity(pData->fElasticity);
+        if (pData->ucFlags & 0x400)
+            pObject->SetBuoyancyConstant(pData->fBuoyancyConstant);
+        if (pData->ucFlags & 0x800)
+            pObject->SetCenterOfMass(pData->vecCenterOfMass);
+
+        // We check whether the object is marked as "moving" and whether it is actually still moving
+        if (pObject->m_bIsMoving && !pObject->IsMoving())
+        {
+            CLuaArguments Arguments;
+            pObject->CallEvent("onObjectMoveStop", Arguments);
 
-        // Tell everyone
-        m_pPlayerManager->BroadcastOnlyJoined(Packet, pPlayer);
+            pObject->m_bIsMoving = false;
+        }
```

Comment:
No. This event is broken by design.
Objects without syncer won't trigger it. Real "stop time" and "trigger time" may differ significantly as a result of this.

Reply:
This is true, so an appropriate note should be added to the wiki about it (if it ever gets merged).

Events such as onPedDamage, onPedWasted, etc. will also not be triggered without a syncer. If you place a ped next to a burning vehicle and leave the server, the ped damage event will not be triggered

Reply:
> If you place a ped next to a burning vehicle and leave the server, the ped damage event will not be triggered

But a ped doesn't deal damage without a syncer, and an object stops without a syncer. Why do you need API with such bad behavior? I see no reason to use this event instead of setTimer.

Reply:
The server does not check whether the object has reached its destination, this happens on the client side. So you can't clearly say that an object stops on an empty server, because no object physically exists on an empty server. Its status is set at the time of joining the server in CEntityAddPacket.

Reply:
> So you can't clearly say that an object stops on an empty server

You can check `IsMoving`, so server knows when an object moving and when it stops

Reply:
@TheNormalnij What do you think about the new logic? https://github.com/multitheftauto/mtasa-blue/pull/3405/commits/843af2d030042c413f6f8d855396fccf8359fd1c

Reply:
Yeah, it's better